src/java.base/share/classes/sun/nio/ch/NativeDispatcher.java

```diff
@@ -69,11 +72,28 @@ abstract long writev(FileDescriptor fd, long address, int len)
 
     abstract void close(FileDescriptor fd) throws IOException;
 
-    // Prepare the given fd for closing by duping it to a known internal fd
-    // that's already closed.  This is necessary on some operating systems
-    // (Solaris and Linux) to prevent fd recycling.
-    //
-    void preClose(FileDescriptor fd) throws IOException {
+    /**
+     * Prepare the given file descriptor for closing. If a virtual thread is blocked
+     * on the file descriptor then it is unparked so that it stops polling. On Unix systems,
+     * if a platform thread is blocked on the file descriptor then the file descriptor is
+     * dup'ed to a special fd and the thread signalled so that the syscall fails with EINTR.
+     */
+    final void preClose(FileDescriptor fd, long reader, long writer) throws IOException {
+        if (NativeThread.isVirtualThread(reader) || NativeThread.isVirtualThread(writer)) {
+            int fdVal = JIOFDA.get(fd);
```

Comment:
Hello Alan, previously, in the case of `DatagramChannelImpl`, the `fdVal` was a `final` that was determined when the channel was constructed. Now, that `fdVal` gets determined as and when needed and I see that `FileDescriptor` allows this underlying value to be changed/reset. Is there anything practical to consider here in context of a `fdVal` that has been reset in the `FileDescriptor`?

Reply:
repairSocket doesn't change fd/fdVal, instead they will be connected to a new socket if the method succeeds (or the old socket if it fails).

Reply:
It was the `FileDescriptor.set(int fd)` method which I had in mind. But I now looked at its references and I see that there aren't any in the code paths that this PR deals with. So I believe the proposed change looks good.